internal/storage/recovery_points.go

```diff
@@ -47,39 +47,110 @@ func (s *Storage) readLastCommitContext() (*CommitContext, error) {
 	return &cc, nil
 }
 
-func (s *Storage) readLogEntryBoundaries() (first, last *varlogpb.LogSequenceNumber, err error) {
-	it := s.commitDB.NewIter(&pebble.IterOptions{
+func (s *Storage) readLogEntryBoundaries() (first, last *varlogpb.LogSequenceNumber) {
+	dit := s.dataDB.NewIter(&pebble.IterOptions{
+		LowerBound: []byte{dataKeyPrefix},
+		UpperBound: []byte{dataKeySentinelPrefix},
+	})
+	cit := s.commitDB.NewIter(&pebble.IterOptions{
 		LowerBound: []byte{commitKeyPrefix},
 		UpperBound: []byte{commitKeySentinelPrefix},
 	})
 	defer func() {
-		_ = it.Close()
+		_ = dit.Close()
+		_ = cit.Close()
 	}()
 
-	if !it.First() {
-		return nil, nil, nil
+	first = s.getFirstLogSequenceNumber(cit, dit)
+	if first == nil {
+		return nil, nil
 	}
 
-	firstGLSN := decodeCommitKey(it.Key())
-	firstLE, err := s.readGLSN(firstGLSN)
-	if err != nil {
-		return nil, nil, err
+	last = s.getLastLogSequenceNumber(cit, dit, first)
+	if last == nil {
+		s.logger.Warn("the last must exist but could not be found.", zap.Stringer("first", first))
```

Comment:
Is this a case that could happen? If it happend, `rp.CommittedLogEntry.Last`[L33] is nil and it means `maybe trimmed`[internal/storagenode/logstream/executor.go:L650]. Are there any side effects?

Reply:
This cannot happen because the first log entry already exists. The first log entry is also the last if only a single log entry exists. Therefore, this `if block` is unnecessary but just added for debugging in weird situations. 

These similar anomalies, including this `if block`, can be fixed through synchronization. But as you said, we need to know whether this replica has been trimmed. I think the global low watermark can be helpful.